solver/app/deps.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/omni-network/omni/lib/ethclient/ethbackend"
+	"github.com/omni-network/omni/lib/log"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// procDeps abstracts dependencies for the event processor allowed simplified testing.
+type procDeps struct {
+	ParseID      func(chainID uint64, log types.Log) ([32]byte, error)
+	GetRequest   func(ctx context.Context, chainID uint64, id [32]byte) (bindings.SolveRequest, bool, error)
+	ShouldReject func(ctx context.Context, chainID uint64, req bindings.SolveRequest) (uint8, bool, error)
+
+	Accept  func(ctx context.Context, chainID uint64, req bindings.SolveRequest) error
+	Reject  func(ctx context.Context, chainID uint64, req bindings.SolveRequest, reason uint8) error
+	Fulfill func(ctx context.Context, chainID uint64, req bindings.SolveRequest) error
+	Claim   func(ctx context.Context, chainID uint64, req bindings.SolveRequest) error
+}
+
+func newClaimer(
+	inboxContracts map[uint64]*bindings.SolveInbox,
+	backends ethbackend.Backends,
+	solverAddr common.Address,
+) func(ctx context.Context, chainID uint64, req bindings.SolveRequest) error {
+	return func(ctx context.Context, chainID uint64, req bindings.SolveRequest) error {
+		inbox, ok := inboxContracts[chainID]
+		if !ok {
+			return errors.New("unknown chain")
+		}
+
+		backend, err := backends.Backend(chainID)
+		if err != nil {
+			return err
+		}
+
+		txOpts, err := backend.BindOpts(ctx, solverAddr)
+		if err != nil {
+			return err
+		}
+
+		tx, err := inbox.Claim(txOpts, req.Id)
+		if err != nil {
+			return errors.Wrap(err, "claim request")
+		} else if _, err := backend.WaitMined(ctx, tx); err != nil {
+			return errors.Wrap(err, "wait mined")
+		}
+
+		return nil
+	}
+}
+
+func newFulfiller(
+	outboxContracts map[uint64]*bindings.SolveOutbox,
+	backends ethbackend.Backends,
+	solverAddr common.Address,
+) func(ctx context.Context, chainID uint64, req bindings.SolveRequest) error {
+	return func(ctx context.Context, chainID uint64, req bindings.SolveRequest) error {
+		outbox, ok := outboxContracts[chainID]
+		if !ok {
+			return errors.New("unknown chain")
+		}
+
+		backend, err := backends.Backend(chainID)
+		if err != nil {
+			return err
+		}
+
+		txOpts, err := backend.BindOpts(ctx, solverAddr)
+		if err != nil {
+			return err
+		}
+
+		if ok, err := outbox.DidFulfill(&bind.CallOpts{Context: ctx}, req.Id, chainID, req.Call); err != nil {
+			return errors.Wrap(err, "did fulfill")
+		} else if ok {
+			log.Info(ctx, "Skipping already fulfilled request", "req_id", req.Id)
+			return nil
+		}
+
+		// TODO(corver): Convert req.Deposits into TokenPreReqs
+		var prereqs []bindings.SolveTokenPrereq
+
+		tx, err := outbox.Fulfill(txOpts, req.Id, chainID, solverAddr, req.Call, prereqs)
+		if err != nil {
+			return errors.Wrap(err, "fulfill request")
+		} else if _, err := backend.WaitMined(ctx, tx); err != nil {
```

Comment:
nit: drop the else?

Reply:
why? I feel it is part of the "fullfil" step, so locally part of the same thing.